src/coreclr/jit/fgdiagnostic.cpp

```diff
@@ -3292,8 +3292,48 @@ void Compiler::fgDebugCheckTypes(GenTree* tree)
                 assert(!"Unexpected small type in IR");
             }
 
+            // Validate binary ops
+            if (node->OperIsBinary())
+            {
+                GenTree* op1 = node->gtGetOp1();
+                GenTree* op2 = node->gtGetOp2();
+
+                if (op1->TypeIs(TYP_BYREF) && op2->TypeIs(TYP_BYREF))
+                {
+                    // Technically, the only case where two byrefs could be combined for a reason is a
+                    // SUB<TYP_I_IMPL> to calculate the distance between them. We don't expect the SUB
+                    // itself to be typed as BYREF. Related: https://github.com/dotnet/runtime/issues/84291
+                    if (!node->OperIs(GT_COMMA, GT_QMARK, GT_COLON) && !node->OperIsStore() && node->TypeIs(TYP_BYREF))
+                    {
+                        m_compiler->gtDispTree(node);
+                        assert(!"Unexpected type of a binary operator on two byrefs");
+                    }
+                }
```

Comment:
Not really a fan of a catch-all like this. Nothing intrinsic says that new binary operators shouldn't be working on byref, so this just becomes another place to update if you add a new binary operator that works with byrefs.

I would rather that we have a switch on the different oper types and then assert specific expectations based on the kind of oper. Like most arithmetic nodes should assert that their operands and target types aren't mismatched, `ADD` should check that `ref + I_IMPL => byref`, etc. Some provisions will be necessary to allow implicit `I_IMPL <-> BYREF` conversions.